File: patterns/python/monotonic_stack.py
class MonotonicStack:
    def next_greater_element(self, nums):
        n = len(nums)
        result = [-1] * n  # Default to -1 if no greater element exists
        stack = []  # Stack stores indices

        for i in range(n):
            while stack and nums[i] > nums[stack[-1]]:
                index = stack.pop()
                result[index] = nums[i]
            stack.append(i)
        
        return result

    def daily_temperatures(self, temperatures):
        n = len(temperatures)
        result = [0] * n  # Result array initialized with 0s
        stack = []  # Monotonic decreasing stack

        for i in range(n):
            while stack and temperatures[i] > temperatures[stack[-1]]:
                prev_index = stack.pop()
                result[prev_index] = i - prev_index
            stack.append(i)

        return result

File: patterns/go/monotonic_stack.go
package main

import "fmt"

func nextGreaterElement(nums []int) []int {
	n := len(nums)
	result := make([]int, n)
	for i := range result {
		result[i] = -1 // Default to -1 if no greater element exists
	}
	stack := []int{} // Stack stores indices

	for i := 0; i < n; i++ {
		for len(stack) > 0 && nums[i] > nums[stack[len(stack)-1]] {
			index := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			result[index] = nums[i]
		}
		stack = append(stack, i)
	}
	return result
}

func dailyTemperatures(temperatures []int) []int {
	n := len(temperatures)
	result := make([]int, n) // Result array initialized with 0s
	stack := []int{}         // Monotonic decreasing stack

	for i := 0; i < n; i++ {
		for len(stack) > 0 && temperatures[i] > temperatures[stack[len(stack)-1]] {
			prevIndex := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			result[prevIndex] = i - prevIndex
		}
		stack = append(stack, i)
	}
	return result
}

func main() {
	nums := []int{2, 1, 5, 6, 2, 3}
	fmt.Println(nextGreaterElement(nums))

	temperatures := []int{73, 74, 75, 71, 69, 72, 76, 73}
	fmt.Println(dailyTemperatures(temperatures))
}
